src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -1558,10 +1723,48 @@ CHIP_ERROR GroupDataProviderImpl::RemoveFabric(chip::FabricIndex fabric_index)
 CHIP_ERROR GroupDataProviderImpl::Decrypt(PacketHeader packetHeader, PayloadHeader & payloadHeader,
                                           System::PacketBufferHandle & msg)
 {
-    (void) packetHeader;
-    (void) payloadHeader;
-    (void) msg;
-    return CHIP_NO_ERROR;
+    chip::GroupId group_id = packetHeader.GetDestinationGroupId().Value();
+
+    FabricList fabric_list;
+    ReturnErrorOnFailure(fabric_list.Load(mStorage));
+
+    // Search the matching index in the fabric list
+    FabricData fabric;
+    fabric.fabric_index = fabric_list.first_fabric;
+    for (size_t i = 0; i < fabric_list.fabric_count; i++, fabric.fabric_index = fabric.next)
+    {
+        ReturnErrorOnFailure(fabric.Load(mStorage));
+
+        // Iterate key sets
+        KeyMapData mapping(fabric.fabric_index, fabric.first_map);
+
+        // Look for the target group in the fabric's keyset-group pairs
+        for (uint16_t j = 0; j < fabric.map_count; ++j, mapping.id = mapping.next)
+        {
+            ReturnErrorOnFailure(mapping.Load(mStorage));
+            if (mapping.group_id == group_id)
+            {
+                // Group found, get the keyset
+                KeySet keyset;
+                ReturnErrorOnFailure(GetKeySet(fabric.fabric_index, mapping.keyset_id, keyset));
+
+                // Get the operational key
+                chip::MessageSecurity::Groups::OperationKeyType operational_key;
+                ByteSpan epoch_key = keyset.GetCurrentKey();
+                if (epoch_key.empty())
+                {
+                    continue;
+                }
+                ReturnErrorOnFailure(chip::MessageSecurity::Groups::DeriveOperationalKey(epoch_key, operational_key));
+                VerifyOrReturnError(CHIP_NO_ERROR !=
+                                        chip::MessageSecurity::Decrypt(ByteSpan(operational_key, sizeof(operational_key)),
```

Comment:
If this fails, shouldn't we go on to the next matching key?  Maybe it failed because we ended up with a group session id collision and this is the wrong key....